File: examples/java-ddd/application/src/main/java/org/myddd/java/application/UserApplicationImpl.java
package org.myddd.java.application;

import org.myddd.java.api.UserApplication;
import org.myddd.java.api.dto.UserDTO;
import org.myddd.java.application.assembler.UserAssembler;
import org.myddd.java.domain.User;
import org.myddd.querychannel.QueryChannelService;
import org.myddd.utils.Page;

import jakarta.inject.Inject;
import jakarta.inject.Named;
import jakarta.transaction.Transactional;
import java.util.stream.Collectors;

@Named
public class UserApplicationImpl implements UserApplication {

    @Inject
    private QueryChannelService queryChannelService;

    @Override
    @Transactional
    public UserDTO createUser(UserDTO userDTO) {
        var user = UserAssembler.toUser(userDTO);
        var createdUser = user.createUser();
        return UserAssembler.toUserDTO(createdUser);
    }

    @Override
    public Page<UserDTO> searchUser(int page, int pageSize, String searchKey) {
        var pageSearch = queryChannelService
                .createJpqlQuery("from User where name like :name", User.class)
                .addParameter("name","%" + searchKey + "%")
                .setPage(page,pageSize)
                .pagedList();
        return Page.builder(UserDTO.class)
                .start(pageSearch.getStart())
                .totalSize(pageSearch.getResultCount())
                .data(pageSearch.getData().stream().map(UserAssembler::toUserDTO).collect(Collectors.toList()));
    }
}

File: samples/kotlin-ddd/application/src/main/kotlin/org/myddd/kotlin/application/UserApplicationImpl.kt
package org.myddd.kotlin.application

import org.myddd.domain.InstanceFactory
import org.myddd.kotlin.api.UserApplication
import org.myddd.kotlin.api.dto.UserDTO
import org.myddd.kotlin.application.assembler.toDTO
import org.myddd.kotlin.application.assembler.toUser
import org.myddd.kotlin.domain.User
import org.myddd.querychannel.QueryChannelService
import org.myddd.utils.Page
import javax.inject.Named
import javax.transaction.Transactional

@Named
open class UserApplicationImpl:UserApplication {

    private val queryChannelService by lazy { InstanceFactory.getInstance(QueryChannelService::class.java) }

    @Transactional
    override fun createUser(userDTO: UserDTO): UserDTO? {
        val user = userDTO.toUser()
        return user.createUser()?.toDTO()
    }

    override fun pageSearchUser(search: String, pageIndex: Int, pageSize: Int): Page<UserDTO> {
        val pageQuery = queryChannelService
            .createJpqlQuery("from User where name like :name",User::class.java)
            .addParameter("name","%$search%")
            .setPage(pageIndex,pageSize)
            .pagedList()

        return Page.builder(UserDTO::class.java)
            .data(pageQuery.data.map { it.toDTO() })
            .totalSize(pageQuery.resultCount)
            .pageSize(pageSize)
            .start(pageIndex.toLong())
    }
}